[BUG] Operator crashes with ISMPolicy

### What is the bug?

A certain ISMPolicy is able to crash the operator, this should not happen.

### How can one reproduce the bug?

Apply this policy

```
apiVersion: opensearch.opster.io/v1
kind: OpenSearchISMPolicy
metadata:
  creationTimestamp: "2024-11-08T12:37:11Z"
  finalizers:
  - opster.io/opensearch-data
  generation: 2
  labels:
    k8slens-edit-resource-version: v1
  name: ism-devops
  namespace: opensearch-testing
  resourceVersion: "369198011"
spec:
  defaultState: hot
  description: A basic policy that will immediately move data to warm, then delete
    after 365 days.
  errorNotification: {}
  ismTemplate:
    indexPatterns:
    - acr-logstash-*
    - github-logstash-*
    - azdevops-logstash-*
    priority: 1
  opensearchCluster:
    name: opensearch-cluster-testing
  policyId: ism_devops
  states:
  - actions: []
    name: hot
    transitions:
    - conditions: {}
      stateName: warm
  - actions:
    - {}
    name: warm
    transitions:
    - conditions: {}
      stateName: cold
  - actions: []
    name: cold
    transitions:
    - conditions: {}
      stateName: delete
  - actions:
    - delete: {}
    name: delete
```

### What is the expected behavior?

Apply this policy and the operator crashes, manually removing the object is needed.

### What is your host/environment?

Kubernetes 1.29.0

### Do you have any screenshots?
```
{"level":"info","ts":"2024-11-08T12:38:31.097Z","msg":"Reconciling OpensearchISMPolicy","controller":"opensearchismpolicy","controllerGroup":"o
{"level":"info","ts":"2024-11-08T12:38:31.097Z","msg":"Reconciling OpensearchIndexTemplate","controller":"opensearchindextemplate","controllerG
{"level":"info","ts":"2024-11-08T12:38:31.097Z","msg":"Starting workers","controller":"opensearchuserrolebinding","controllerGroup":"opensearch
{"level":"info","ts":"2024-11-08T12:38:31.097Z","msg":"Reconciling OpenSearchCluster","controller":"opensearchcluster","controllerGroup":"opens
{"level":"info","ts":"2024-11-08T12:38:31.612Z","msg":"Observed a panic in reconciler: runtime error: invalid memory address or nil pointer der
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x17203ec]

goroutine 423 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:115 +0x1e5
panic({0x192c660?, 0x2c45b60?})
    /usr/local/go/src/runtime/panic.go:770 +0x132
github.com/Opster/opensearch-k8s-operator/opensearch-operator/pkg/reconcilers.(*IsmPolicyReconciler).CreateISMPolicy(0xc0066ab780)
    /workspace/pkg/reconcilers/ismpolicy.go:313 +0x36c
github.com/Opster/opensearch-k8s-operator/opensearch-operator/pkg/reconcilers.(*IsmPolicyReconciler).Reconcile(0xc0066ab780)
    /workspace/pkg/reconcilers/ismpolicy.go:172 +0x8f3
github.com/Opster/opensearch-k8s-operator/opensearch-operator/controllers.(*OpensearchISMPolicyReconciler).Reconcile(0xc0003b1ea0, {0x1e0f790,
    /workspace/controllers/opensearchism_controller.go:53 +0x385
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1e0f790?, {0x1e0f790?, 0xc000fc8210?}, {{{0xc00091a2b8?, 0x185
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:118 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0006a0820, {0x1e0f7c8, 0xc0003b19a0}, {0x19aaae0, 0xc0
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:314 +0x345
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0006a0820, {0x1e0f7c8, 0xc0003b19a0})
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265 +0x1c9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 99
    /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:222 +0x50c
```

### Do you have any additional context?

--- a/opensearch-operator/pkg/reconcilers/ismpolicy.go
+++ b/opensearch-operator/pkg/reconcilers/ismpolicy.go
@@ -171,7 +171,9 @@ func (r *IsmPolicyReconciler) Reconcile() (retResult ctrl.Result, retErr error) 
 
 	newPolicy, retErr := r.CreateISMPolicy()
 	if retErr != nil {
-		r.logger.Error(retErr, reason)
+		shortReason := "failed to generate ism policy document"
+		reason = fmt.Sprintf("%s: %s", shortReason, retErr.Error())
+		r.logger.Error(retErr, shortReason)
 		return ctrl.Result{
 			Requeue:      true,
 			RequeueAfter: defaultRequeueAfter,
@@ -186,9 +188,10 @@ func (r *IsmPolicyReconciler) Reconcile() (retResult ctrl.Result, retErr error) 
 		}
 		retErr = services.CreateISMPolicy(r.ctx, r.osClient, request, policyId)
 		if retErr != nil {
-			reason = "failed to create ism policy"
-			r.logger.Error(retErr, reason)
-			r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+			shortReason := "failed to create ism policy"
+			reason = fmt.Sprintf("%s: %s", shortReason, retErr.Error())
+			r.logger.Error(retErr, shortReason)
+			r.recorder.Event(r.instance, "Warning", opensearchAPIError, shortReason)
 			return ctrl.Result{
 				Requeue:      true,
 				RequeueAfter: defaultRequeueAfter,
@@ -262,9 +265,10 @@ func (r *IsmPolicyReconciler) Reconcile() (retResult ctrl.Result, retErr error) 
 	}
 	retErr = services.UpdateISMPolicy(r.ctx, r.osClient, request, &existingPolicy.SequenceNumber, &existingPolicy.PrimaryTerm, existingPolicy.PolicyID)
 	if retErr != nil {
-		reason = "failed to update ism policy with Opensearch API"
-		r.logger.Error(retErr, reason)
-		r.recorder.Event(r.instance, "Warning", opensearchAPIError, reason)
+		shortReason := "failed to update ism policy with Opensearch API"
+		reason = fmt.Sprintf("%s: %s", shortReason, retErr.Error())
+		r.logger.Error(retErr, shortReason)
+		r.recorder.Event(r.instance, "Warning", opensearchAPIError, shortReason)
 		return ctrl.Result{
 			Requeue:      true,
 			RequeueAfter: defaultRequeueAfter,
@@ -283,34 +287,37 @@ func (r *IsmPolicyReconciler) CreateISMPolicy() (*requests.ISMPolicySpec, error)
 		DefaultState: r.instance.Spec.DefaultState,
 		Description:  r.instance.Spec.Description,
 	}
-	if r.instance.Spec.ErrorNotification != nil {
+	if r.instance.Spec.ErrorNotification != nil && r.instance.Spec.ErrorNotification.Destination != nil && r.instance.Spec.ErrorNotification.MessageTemplate != nil {
 		dest := requests.Destination{}
-		if r.instance.Spec.ErrorNotification.Destination != nil {
-			if r.instance.Spec.ErrorNotification.Destination.Amazon != nil {
-				dest.Amazon = &requests.DestinationURL{
-					URL: r.instance.Spec.ErrorNotification.Destination.Amazon.URL,
-				}
-			}
-			if r.instance.Spec.ErrorNotification.Destination.Chime != nil {
-				dest.Chime = &requests.DestinationURL{
-					URL: r.instance.Spec.ErrorNotification.Destination.Chime.URL,
-				}
-			}
-			if r.instance.Spec.ErrorNotification.Destination.Slack != nil {
-				dest.Slack = &requests.DestinationURL{
-					URL: r.instance.Spec.ErrorNotification.Destination.Slack.URL,
-				}
-			}
-			if r.instance.Spec.ErrorNotification.Destination.CustomWebhook != nil {
-				dest.CustomWebhook = &requests.DestinationURL{
-					URL: r.instance.Spec.ErrorNotification.Destination.CustomWebhook.URL,
-				}
+		if r.instance.Spec.ErrorNotification.Destination.Amazon != nil {
+			dest.Amazon = &requests.DestinationURL{
+				URL: r.instance.Spec.ErrorNotification.Destination.Amazon.URL,
 			}
 		}
+		if r.instance.Spec.ErrorNotification.Destination.Chime != nil {
+			dest.Chime = &requests.DestinationURL{
+				URL: r.instance.Spec.ErrorNotification.Destination.Chime.URL,
+			}
+		}
+		if r.instance.Spec.ErrorNotification.Destination.Slack != nil {
+			dest.Slack = &requests.DestinationURL{
+				URL: r.instance.Spec.ErrorNotification.Destination.Slack.URL,
+			}
+		}
+		if r.instance.Spec.ErrorNotification.Destination.CustomWebhook != nil {
+			dest.CustomWebhook = &requests.DestinationURL{
+				URL: r.instance.Spec.ErrorNotification.Destination.CustomWebhook.URL,
+			}
+		}
+		if dest.Amazon == nil && dest.Chime == nil && dest.Slack == nil && dest.CustomWebhook == nil {
+			return nil, errors.New("exactly one errorNotification.destination must be set")
+		}
+		messageTemplate := requests.MessageTemplate{Source: r.instance.Spec.ErrorNotification.MessageTemplate.Source}
+
 		policy.ErrorNotification = &requests.ErrorNotification{
 			Channel:         r.instance.Spec.ErrorNotification.Channel,
 			Destination:     &dest,
-			MessageTemplate: &requests.MessageTemplate{Source: r.instance.Spec.ErrorNotification.MessageTemplate.Source},
+			MessageTemplate: &messageTemplate,
 		}
 	}
 

